Initial commit for versioning ADR (#3210)
conveyordata/data-product-portal: docs/adr/0015-versioning.md

* **Option 3: Full versioning** - Every change (major/minor/patch) creates new Output Port + new Technical Asset

## Decision Outcome

**Chosen option: Option 2 — Branch-based versioning.**

Version only when necessary: when a change would break existing Consumers. This means major version bumps only (v1 → v2 → v3).

**Rationale:**

The fundamental insight is that **data products evolve differently than software.** In software, you can have many minor versions (v1.0, v1.1, v1.2) because each represents a stable release at a point in time. In data products, there is no "release" - data flows continuously.

Trying to version every schema change creates an unsustainable operational burden. A Producer would need to maintain views for v1.0, v1.1, v1.2, v1.3, etc., all reading from the same evolving table. This doesn't match how databases actually work in practice.

Instead, think of versions as "branches":
- **v1**: The current stable version, continuously evolving with backward-compatible changes
- **v2**: A new branch created when a breaking change is necessary
- Old versions (v1) remain frozen or continue with backported fixes only

> This aligns with how APIs are versioned in practice (`/api/v1/`, `/api/v2/`) and how databases are actually managed.

Reviewer: I understand what you are trying to say, only having a change for a major version has impact and thus needs to create new technical assets etc, and fundamentally changes your output port.

But that doesn't mean there is a value in having a changelog for your v1, just because only the latest version of v1 available doesn't mean you don't want to tell people a new field is available. Or that certain fields are deprecated in favor of something else.

There is value in that.

Now I do agree the MVP is just supporting major versions

### How it works

**Backward-compatible changes** (no new version):
- Producer updates the Technical Asset in place
- Examples: Adding nullable columns, adding new API endpoints, new S3 files
- All Consumers on v1 automatically see the changes

> - Document in changelog, but no migration required

Reviewer: I notice you add a changelog here, why not just bump the version number then :) From v1 to v1.1

**Breaking changes** (major version bump):
- Producer creates new Output Port (v2) with new Technical Asset
- Examples: Removing columns, renaming columns, changing data types, schema restructuring
- Consumers on v1 continue using old infrastructure
- Consumers explicitly request access to v2 when ready to migrate
- Both v1 and v2 run in parallel during migration period

**When to version:** Only when the change would break existing Consumer queries or code.

**Key principle:** Versions are long-lived branches, not frequent releases. A data product might have only 2-3 versions over its lifetime (v1, v2, maybe v3).

**Important:** Each Output Port version MUST have its own Technical Asset. You cannot have v1 and v2 pointing to the same table/bucket/API - if they did, they would return identical data and versioning would be meaningless. This means versioning always creates infrastructure duplication.

### Confirmation

- Output Ports have an optional `version` field (default: v1)

> - Version is a simple integer: v1, v2, v3 (not semantic versioning)

Reviewer: I disagree with this, there is still value in having semantic versioning, however only a major version change results in need to create a new technical asset.
Agree that this is an MVP though

- Creating a new version creates a new Output Port record + new Technical Asset
- Both versions coexist until the old version is explicitly deprecated
- Handshakes reference a specific Output Port version
- Backward-compatible changes are made in-place to existing versions
- Changelog tracks all changes (versioned or not)

## Pros and Cons of the Options

### Option 1: No versioning

* **Good, because** simplest possible solution - no new infrastructure needed
* **Good, because** matches how most internal databases work today
* **Good, because** no operational overhead of maintaining multiple versions
* **Bad, because** Producers can't make breaking changes without breaking Consumers
* **Bad, because** forces workaround of creating separate Output ports and technical assets for v1/v2
* **Bad, because** versioning is not managed by portal and multiple versions show up as duplicates in the marketplace

### Option 2: Branch-based versioning (major versions only)

* **Good, because** operationally realistic - only version when necessary

> * **Good, because** matches how APIs are versioned in practice (v1, v2, v3)

Reviewer: API's still have minor and bugfix versions :) it's just that the v1 version always uses the latest v1, and you can't go back to a previous minor of v1

* **Good, because** aligns with how databases evolve (backward-compatible schema additions are normal)
* **Neutral, because** requires Producers to understand what "breaking" means for their Consumers
* **Bad, because** backward-compatible changes happen without Consumer opt-in

### Option 3: Full versioning (all changes create new versions)

* **Good, because** maximum Consumer control - every change is opt-in
* **Good, because** clean separation - each version is completely independent
* **Bad, because** operationally unsustainable - would require maintaining dozens of views/endpoints
* **Bad, because** doesn't match how databases actually evolve
* **Bad, because** creates massive infrastructure overhead
* **Bad, because** unclear when to version (is adding an index a new version?)

---

## Design Details

### What constitutes a breaking change?

Changes that would cause existing Consumer queries or code to fail:

**Breaking (requires new version):**
- Removing a column
- Renaming a column
- Changing a column's data type
- Removing an API endpoint
- Changing API response structure
- Restructuring file formats
- Removing required fields

**Not breaking (update in-place):**

> - Adding a new nullable column

Reviewer: ```suggestion
- Adding a new column
```

- Adding a new API endpoint

> - Adding optional fields to JSON

Reviewer: ```suggestion
- Adding fields to JSON
```

- Adding new files to S3 prefix
- Performance improvements
- Data quality fixes
- Documentation updates

**Gray area (Producer decides):**

> - Adding a new non-nullable column (breaks `SELECT *` but not explicit column lists)

Reviewer: This is non breakable no?

> - Changing column order (breaks `SELECT *`)

Reviewer: This is non breaking

> - Changing data freshness (e.g., real-time → daily batch)

Reviewer: This is breaking

4. Both versions run in parallel
5. Consumers migrate at their own pace
6. Old version is eventually deprecated and removed

### Example: Breaking change (removing a column)

**Initial state:**
- Output Port "Customer Data" v1 → Table `customers` (id, name, email, phone)

**Change needed:**
- Remove `phone` column for privacy compliance

**Steps:**
1. Create new table: `customers_v2` (id, name, email)
2. Create Output Port "Customer Data" v2 → points to `customers_v2`
3. Set up data sync from `customers` to `customers_v2` (minus phone column)
4. Consumers on v1 continue using `customers` table
5. New Consumers or migrated Consumers use v2 → `customers_v2`
6. After migration period, deprecate v1

### Example: Non-breaking change (adding a column)

**Initial state:**
- Output Port "Customer Data" v1 → Table `customers` (id, name, email)

**Change needed:**
- Add `loyalty_tier` column

**Steps:**
1. Add column to existing table: `ALTER TABLE customers ADD COLUMN loyalty_tier VARCHAR(20)`
2. Update Output Port metadata (schema documentation)
3. No new version needed - all Consumers on v1 automatically have access to new column
4. Consumers can choose to use new column or ignore it

> ### Database schema

Reviewer: Is this also an example?

```sql
-- Add version to Output Ports (datasets)
ALTER TABLE datasets ADD COLUMN version INTEGER DEFAULT 1;
ALTER TABLE datasets ADD COLUMN parent_version_id UUID REFERENCES datasets(id);

-- Add unique constraint: can't have duplicate versions for same logical port
CREATE UNIQUE INDEX idx_datasets_version
ON datasets(name, data_product_id, version);
```

No version states (active/deprecated/sunset) needed - just delete old versions when fully migrated.

> ## Implications and Trade-offs

Reviewer: I think we can remove thsi section, it's a bit more of the same

### What this doesn't solve

- Consumers still auto-upgraded for non-breaking changes
- Requires Producers to correctly identify what's "breaking"
- Requires infrastructure duplication during migration periods
- No automatic rollback if a change breaks Consumers

### Comparison with software versioning

This is intentionally different from semantic versioning (major.minor.patch):

**Software versioning:**
- Many releases (v1.0.0, v1.1.0, v1.2.0, v2.0.0)
- Each release is a snapshot in time
- Clear "release" moment

**Data product versioning:**
- Few versions (v1, v2, maybe v3)
- Each version is a living thing that evolves
- No "release" - data flows continuously

Trying to apply software versioning to data products leads to operational nightmares.

---

## Open Questions

1. **What if Consumers disagree on whether a change is breaking?**
   - A Producer thinks adding a column is non-breaking (it's nullable)
   - But a Consumer using `SELECT *` will get broken code
   - Decision: Producer determines breaking-ness, but should be conservative
   - Recommendation: Producers should avoid `SELECT *` in their documentation

> 

Reviewer: Does this break? It won't in spark or pandas for example.

2. **How long should old versions be supported?**
   - No technical enforcement - purely a governance/SLA question
   - Recommendation: Set deprecation timeline per Data Product (e.g., 6 months notice)

>    - Portal can track which Consumers are on which version to inform deprecation

Reviewer: I think we should support setting deprecation timeline when making a major version bump

3. **What about data contracts?**
   - If data contracts are adopted, they could help identify breaking changes
   - Contract diff tool could flag schema changes as breaking/non-breaking
   - This is complementary to versioning, not a replacement

>    - See: ADR-0012 (BiToL standard supports data contracts)

Reviewer: Each major version has a differnt contract no?

> 4. **Should there be a "latest" alias?**

Reviewer: Nope I would encourage to not do this, also we are not a technical catalog so what does it even mean?

When people consume an output port they consume every version anyway no?

   - Deferred: Start without aliases
   - Forces Consumers to make explicit version choices
   - Can add later if there's demand for "always use latest stable"

> 5. **How does this interact with data quality issues?**

Reviewer: I guess data quality should be posted per major version?

- v1 and v2 coexist until v1 is deprecated.

This gives both change tracking (via semver) and operational realism (infra duplication only when needed).

### Why semver?

Data pipelines have CI/CD release cycles just like regular software. A dbt project is committed, reviewed, and deployed. Each deployment that changes the schema IS a release. The distinction is: version on **code deploy** (schema change), not on every pipeline **run** (data refresh). This is no different from an API that versions on deploy, not on every request.

### Why only branch on major?

A minor version (adding a column) is backward-compatible by definition - the Consumer already has access to the new data. Creating separate infrastructure for each minor version leads to operational sprawl (maintaining v1.0, v1.1, v1.2, v1.3 views all pointing to the same underlying table). What Consumers need for minor changes is **notification**, not a separate version with its own Technical Asset.

### How it works

**Patch changes** (1.0.0 → 1.0.1):
- Data quality fixes, documentation updates, performance improvements
- Update in place, bump version number on existing Output Port
- Notify Consumers via changelog

> **Minor changes** (1.0.0 → 1.1.0):
> - Backward-compatible additions (new column, new file, new API field)
> - Update in place, bump version number on existing Output Port
> - Notify Consumers via changelog

Reviewer: ```suggestion
**Minor changes** (1.0.0 → 1.1.0):
- Backward-compatible additions (new column, new file, new API field)
- Update in place, bump version number on existing Output Port
- Notify Consumers via changelog
- deprecations, warn users of fields that will be removed in a new version, and give them an alternative in the changelog
```

**Breaking changes** (1.x.x → 2.0.0):
- Producer creates new Output Port version (v2) with new Technical Asset (or existing one in case of schemas)
- Examples: Removing columns, renaming columns, changing data types, schema restructuring
- Consumers on v1 continue using old infrastructure
- Access to v2 is **automatically inherited** — no re-requesting
- Consumers confirm when they're ready to switch; only after can the v1 be removed.
- Both versions coexist during migration period

**When to version:** Only when the change would break existing Consumer queries or code.

**Key principle:** Versions are long-lived branches, not frequent releases. A data product might have only 2-3 major versions over its lifetime.

### Access inheritance for major versions

> When a Producer creates v2 of an Output Port, all Consumers with active Handshakes on v1 automatically receive access to v2. They don't need to go through the access request flow again - requiring re-requests would recreate the exact problem we're solving today.

Reviewer: ```suggestion
When a Producer creates v2 of an Output Port, all Consumers with an input port on v1 automatically receive access to v2. They don't need to go through the access request flow again - requiring re-requests would recreate the exact problem we're solving today.
```

### Example: Breaking change with schema-level Technical Asset (dbt)

**Initial state:**
- Output Port "Marketing Gold" v1.2.0 → Technical Asset: Schema `gold.marketing`
- Schema contains: `customers`, `orders`, `products`, and 7 other tables

**Change needed:**
- Restructure the `customers` table (rename columns)

**Steps:**
1. dbt creates `customers_v2` model in the same `gold.marketing` schema
2. Create Output Port "Marketing Gold" v2.0.0 → **same** Technical Asset (schema `gold.marketing`)
3. v2 contract documents: "use `customers_v2` instead of `customers`"
4. All v1 Consumers automatically get Handshakes on v2
5. Consumers update their queries (only the `customers` reference changes) and confirm
6. The other 9 tables are unaffected — no new Technical Assets needed

### Example: Non-breaking change (adding a column)

**Initial state:**
- Output Port "Customer Data" v1.2.0 → Table `customers` (id, name, email)

**Change needed:**
- Add `loyalty_tier` column

**Steps:**
1. Add column to existing table: `ALTER TABLE customers ADD COLUMN loyalty_tier VARCHAR(20)`
2. Bump version to 1.3.0 in Portal, add changelog entry: "Added optional loyalty_tier column"

> 3. Consumers are notified; no migration required - the new column is already available

Reviewer: I think this is redundant, and can be removed for readability
